### PR TITLE
[expo-updates][iOS] Logging and log reading

### PR DESCRIPTION
**Pull request by douglowder**
# Why

Provides a way to log from both Objective-C and Swift code in expo-updates, and provides code to read all logs written since a given time.

# How

- New Swift code implementing the logger, log reader, expo-updates specific error codes, and JSON format for log messages
- Logs with severity "error" or "fatal" will include stack traces (limited to 20 frames)
- Leverages the existing Logger in expo-modules-core.

# Test Plan

New unit tests provided for both Swift and Objective-C.

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [x] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+      // Fetch log objects.
+      let allEntries = try logStore.getEntries(at: position)
+
+      // Filter the log to select our subsystem and the expo-updates category
+      // and extract just the log message strings
+      let allEntriesMessages = allEntries
+          .compactMap { $0 as? OSLogEntryLog }
+          .filter { $0.subsystem == "dev.expo.modules" }
+          .filter { $0.category == "expo-updates" }
+          .compactMap { $0.composedMessage }
```
Comment (esamelson) [inline packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift:36]: It looks like we're reading all entries into memory and _then_ filtering them; do you think performance is something worth being concerned about here? (Should we put an upper bound on the `newerThan` date, for example?)

File: packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogEntry.swift [hunk 2]
```diff
@@ -0,0 +1,43 @@
+// Copyright 2022-present 650 Industries. All rights reserved.
+
+// Schema for the fields in expo-updates log message JSON strings
+
+import Foundation
+
+public struct UpdatesLogEntry: Codable {
```
Comment (esamelson) [inline packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogEntry.swift:7]: is it possible to expose this to obj-c if it also subclasses NSObject? Seems like it would be nice if the class calling `getLogEntries` weren't also responsible for parsing the entries, but also understandable if it's too tricky to implement across both languages.

File: packages/expo-updates/ios/EXUpdates/Logging/UpdatesErrorCode.swift [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+import Foundation
+
+@objc(EXUpdatesErrorCode)
+public enum UpdatesErrorCode: Int {
+  case None = 0
+  case NoUpdatesAvailable = 1
+  case UpdateAssetsNotAvailable = 2
+  case UpdateServerUnreachable = 3
+  case UpdateHasIncorrectHash = 4
```
Comment (esamelson) [inline packages/expo-updates/ios/EXUpdates/Logging/UpdatesErrorCode.swift:13]: it would probably be more terminologically correct to say `UpdateHasInvalidSignature` as we don't have "hashes" at the update level

**Review by esamelson — COMMENTED** [review]
Cool, this looks like it will be nice to use! I think there will probably be more error codes to add over time -- one that comes to mind is `AssetHasMismatchedHash` or something to that effect -- and it's possible we might not end up using all the ones there now, but probably fine to change/grow the list as needed.

File: packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogEntry.swift (hunk 2 above)
Comment (douglowder) [reply packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogEntry.swift:7]: Sure I can have a look

Comment (douglowder) [reply packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogEntry.swift:7]: Exposing this to Objective C would be a complete reimplementation and be a bit awkward. It might be better to change the reader to return a list of `NSDictionary` objects for Objective-C usage.

File: packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift (hunk 1 above)
Comment (esamelson) [reply packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift:36]: I was thinking more along the lines of enforcing a limit in this method, just to keep us from shooting ourselves in the foot in the future? But if you would rather keep the flexibility (or just don't think it will be an issue) that's totally fine w me too

File: packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogEntry.swift (hunk 2 above)
Comment (esamelson) [reply packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogEntry.swift:7]: Cool -- that makes sense to me. Thanks for checking that out.

File: packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift (hunk 1 above)
Comment (douglowder) [reply packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift]: I did two things here:
- only allow logs to be searched going back 1 day
- added a predicate to the OSLogStore call so we only retrieve the entries we want and don't have to filter after the fact

File: packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogEntry.swift (hunk 2 above)
Comment (douglowder) [reply packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogEntry.swift:7]: Now Objective-C can call the reader and get entries returned as NSDictionaries.

File: packages/expo-updates/ios/EXUpdates/Logging/UpdatesErrorCode.swift (hunk 3 above)
Comment (douglowder) [reply packages/expo-updates/ios/EXUpdates/Logging/UpdatesErrorCode.swift]: Fixed

**Review by esamelson — APPROVED** [review]
LGTM, excited to see this in action!

File: packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogger.swift [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+@available(macOS 12.0, iOS 15.0, watchOS 8.0, tvOS 15.0, *)
+@objc(EXUpdatesLogger)
+public class UpdatesLogger: NSObject {
+
+  private let osLogger: os.Logger
+
+  public override init() {
+    osLogger = os.Logger(subsystem: "dev.expo.modules", category: "expo-updates")
+  }
```
Comment (wschurman) [inline packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogger.swift:18]:
```suggestion
  private let osLogger = os.Logger(subsystem: "dev.expo.modules", category: "expo-updates")
```

File: packages/expo-updates/ios/Tests/EXUpdatesLoggerOCTests.m [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+/*
+#import <XCTest/XCTest.h>
+
+@interface EXUpdatesLoggerOCTests : XCTestCase
+
+@end
+
+// Headers for the Swift classes and enum
+// In application code, EXUpdates-Swift.h can be imported instead
```
Comment (wschurman) [inline packages/expo-updates/ios/Tests/EXUpdatesLoggerOCTests.m:17]: I wonder why this isn't visible to tests. It's been a few months since I did swift stuff so I don't remember if I ever got this bridge working.

That being said, I think just having the swift test is sufficient. We can rely upon the compiler to tell us if the bridge is broken.

File: packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+        let json = entry.element as String
+        if returnEntriesAs == "NSString" {
+          result.add(json)
+          continue
+        }
+        let logEntry = UpdatesLogEntry.create(from: json)
+        result.add(logEntry!.asDict())
+      }
+    } catch {
+      result.add("Error occurred in UpdatesLogReader: \(error.localizedDescription)")
```
Comment (wschurman) [inline packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift:58]: not sure this pattern of returning an error string as an arbitrary element will be consumable. Might make more sense to just throw the error?

This also is a bit messy for the dictionary element return type.

File: packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+  }
+
+  /**
+   Get expo-updates logs newer than the given date
+   Returns an Objective-C NSArray of either strings or NSDictionary objects
+   Returned strings are all in the JSON format of UpdatesLogEntry
+   Maximum of one day lookback is allowed
+   */
+  @objc(getLogEntriesNewerThan:returnEntriesAs:)
+  public func getLogEntries(newerThan: Date, returnEntriesAs: String = "NSDictionary") -> NSArray {
```
Comment (wschurman) [inline packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift:27]: Prefer two methods that have more concrete return types than a dynamic return type

**Review by wschurman — CHANGES_REQUESTED** [review]

File: packages/expo-updates/ios/EXUpdates/Logging/UpdatesErrorCode.swift [hunk 8]
```diff
@@ -0,0 +1,42 @@
+// Copyright 2022-present 650 Industries. All rights reserved.
+
+// Error codes for expo-updates logs
```
[the remark this answers is not in the document]
Comment (douglowder) [reply packages/expo-updates/ios/EXUpdates/Logging/UpdatesErrorCode.swift]: Done

File: packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogger.swift (hunk 4 above)
Comment (douglowder) [reply packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogger.swift]: Done

File: packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift (hunk 6 above)
Comment (douglowder) [reply packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift]: Hmm I will think more on how to handle this better. No error should ever be thrown really, but I'll have a look at this.

File: packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+
+import ExpoModulesCore
+import SafariServices
+
+@available(macOS 12.0, iOS 15.0, watchOS 8.0, tvOS 15.0, *)
+@objc(EXUpdatesLogReader)
+public class UpdatesLogReader: NSObject {
+
+  @objc(getLogEntriesNewerThan:)
+  public func getLogEntries(newerThan: Date) -> NSArray {
```
[the remark this answers is not in the document]
Comment (douglowder) [reply packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift]: Done

File: packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift (hunk 7 above)
Comment (douglowder) [reply packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift]: Done

File: packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogEntry.swift [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+    do {
+      let jsonEncoder = JSONEncoder()
+      let jsonData = try jsonEncoder.encode(self)
+      return String(data: jsonData, encoding: .utf8) ?? nil
+    } catch {
+      return nil
+    }
+  }
+
+  public func asDict() -> NSDictionary {
```
[the remark this answers is not in the document]
Comment (douglowder) [reply packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogEntry.swift]: Done

**Review by tsapeta — CHANGES_REQUESTED** [review]
I quickly went through the PR so it's just the first iteration

File: packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+
+import ExpoModulesCore
+import SafariServices
+
+/**
+ Class to read expo-updates logs using OSLogReader
+ */
+@available(macOS 12.0, iOS 15.0, watchOS 8.0, tvOS 15.0, *)
+@objc(EXUpdatesLogReader)
+public class UpdatesLogReader: NSObject {
```
[the remark this answers is not in the document]
Comment (douglowder) [reply packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift:14]: As written, this class is quite specific to `expo-updates`. However, I agree that in future we could split the more general parts of this class out to something more generic that lives in `expo-modules-core`.

File: packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+    let strings: [String] = getLogEntries(newerThan: newerThan)
+    var result: [[String: Any]] = []
+    strings.forEach { s in
+      let entry = UpdatesLogEntry.create(from: s)
+      if (entry != nil) {
+        let dict = entry?.asDict()
+        if (dict != nil) {
+          result.append(dict!)
+        }
+      }
```
[the remark this answers is not in the document]
Comment (douglowder) [reply packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift:27]: Good point about `compactMap`, I'll take a look at that method

File: packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+    } catch {
+      result.append("Error occurred in UpdatesLogReader: \(error.localizedDescription)")
+    }
+    return result
+  }
+  
+  /**
+   Convenience method for returning expo-updates logs from last hour
+   */
+  @objc public func logEntriesInLastHour() -> [String] {
```
[the remark this answers is not in the document]
Comment (douglowder) [reply packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift:73]: No they were intended to be public. I'll think about whether we still need these or not, though.

File: packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogger.swift [hunk 14]
```diff
@@ -0,0 +1,177 @@
...
+
+import Foundation
+import os.log
+
+import ExpoModulesCore
+
+/**
+ Subclass of Logger that overrides the format of log messages
+ */
+internal class CoreLogger: ExpoModulesCore.Logger {
```
[the remark this answers is not in the document]
Comment (douglowder) [reply packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogger.swift:11]: The intent of this feature was to log messages in JSON format. What I can do instead is remove this change, and handle the prefix in `UpdatesLogReader` during message extraction.

File: packages/expo-modules-core/ios/Swift/Logging/Logger.swift [hunk 15]
```diff
@@ -2,9 +2,13 @@
 
 import Dispatch
 
-public let log = Logger(category: "expo")
+public let log = Logger(category: Logger.LOG_CATEGORY)
+
+open class Logger {
+
+  public static let LOG_SUBSYSTEM = "dev.expo.modules"
+  public static let LOG_CATEGORY = "expo"
```
[the remark this answers is not in the document]
Comment (douglowder) [reply packages/expo-modules-core/ios/Swift/Logging/Logger.swift:10]: Good point, I'll make this change

File: packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift (hunk 12 above)
[the remark this answers is not in the document]
Comment (douglowder) [reply packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift:27]: Fixed

File: packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+   Maximum of one day lookback is allowed
+   */
+  @objc(getLogEntryStringsNewerThan:)
+  public func getLogEntries(newerThan: Date) -> [String] {
+    var result: [String] = []
+    do {
+      let earliestDate = Date().addingTimeInterval(-86400)
+      let dateToUse = newerThan.timeIntervalSince1970 < earliestDate.timeIntervalSince1970 ?
+      earliestDate :
+      newerThan
```
[the remark this answers is not in the document]
Comment (douglowder) [reply packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift:45]: Fixed

File: packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+   Convenience method for returning expo-updates logs from last hour
+   */
+  @objc public func logEntriesInLastHour() -> [String] {
+    return getLogEntries(newerThan: Date().addingTimeInterval(-3600))
+  }
+
+  /**
+   Convenience method for returning expo-updates logs from last day
+   */
+  @objc public func logEntriesInLastDay() -> [String] {
```
[the remark this answers is not in the document]
Comment (douglowder) [reply packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift:80]: Removed this

File: packages/expo-modules-core/ios/Swift/Logging/Logger.swift (hunk 15 above)
[the remark this answers is not in the document]
Comment (douglowder) [reply packages/expo-modules-core/ios/Swift/Logging/Logger.swift:10]: Fixed

File: packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogger.swift (hunk 14 above)
[the remark this answers is not in the document]
Comment (douglowder) [reply packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogger.swift:11]: Fixed

File: packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+   */
+  @objc public func logEntriesInLastHour() -> NSArray {
+    return getLogEntries(newerThan: Date().addingTimeInterval(-3600))
+  }
+
+  /**
+   Convenience method for returning expo-updates logs from last day
+   */
+  @objc public func logEntriesInLastDay() -> NSArray {
+    return getLogEntries(newerThan: Date().addingTimeInterval(-86400))
```
[the remark this answers is not in the document]
Comment (douglowder) [reply packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift]: I'm removing these as I don't think they are really needed

File: packages/expo-updates/ios/Tests/EXUpdatesLoggerOCTests.m (hunk 5 above)
Comment (douglowder) [reply packages/expo-updates/ios/Tests/EXUpdatesLoggerOCTests.m]: Replaced the OC test with a Swift test to verify the dictionary reader method

File: packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift (hunk 6 above)
Comment (douglowder) [reply packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift]: Completed this change

File: packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+
+      // Fetch log objects, selecting our subsystem and category
+      let predicate = NSPredicate(format: "category == %@ AND subsystem = %@",
+                                  argumentArray: [UpdatesLogger.EXPO_UPDATES_LOG_CATEGORY, Logger.EXPO_MODULES_LOG_SUBSYSTEM])
+      let allEntries = try logStore.getEntries(at: position, matching: predicate)
+
+      // Extract just the log message strings, removing the first two characters added
+      // by ExpoModulesCore.Logger
+      result = allEntries
+          .compactMap { $0 as? OSLogEntryLog }
```
Comment (wschurman) [inline packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift:54]: remove

File: packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift [hunk 20]
```diff
@@ -0,0 +1,63 @@
...
+          .compactMap { String($0.composedMessage.suffix(from: $0.composedMessage.index($0.composedMessage.startIndex, offsetBy: 2)))
```
Comment (wschurman) [inline packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift:55]: this can just be `map`

File: packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogEntry.swift [hunk 21]
```diff
@@ -0,0 +1,78 @@
...
+
+  /**
+   Returns a new UpdatesLogEntry from a JSON string, or nil if a decoding error occurs
+   */
+  public static func create(from: String) -> UpdatesLogEntry? {
+    do {
+      let jsonDecoder = JSONDecoder()
+      guard let jsonData = from.data(using: .utf8) else { return nil }
+      let logEntry: UpdatesLogEntry = try jsonDecoder.decode(UpdatesLogEntry.self, from: jsonData)
+      return logEntry
```
Comment (wschurman) [inline packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogEntry.swift:61]:
```suggestion
      return try jsonDecoder.decode(UpdatesLogEntry.self, from: jsonData)
```

**Review by wschurman — APPROVED** [review]
Just a few more changes before commit. May be worth discussing the error case mentioned in the inline comment.

File: packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift (hunk 19 above)
Comment (douglowder) [reply packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift:54]: Done

File: packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift (hunk 20 above)
Comment (douglowder) [reply packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift:55]: I'd prefer to leave it as `compactMap`

File: packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift [hunk 22]
```diff
@@ -0,0 +1,63 @@
+// Copyright 2022-present 650 Industries. All rights reserved.
+
+import Foundation
+import OSLog
+
+import ExpoModulesCore
+import SafariServices
```
[the remark this answers is not in the document]
Comment (douglowder) [reply packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift:7]: Done

File: packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift [hunk 23]
```diff
@@ -0,0 +1,63 @@
...
+      let allEntries = try logStore.getEntries(at: position, matching: predicate)
+
+      // Extract just the log message strings, removing the first two characters added
+      // by ExpoModulesCore.Logger
+      result = allEntries
+          .compactMap { $0 as? OSLogEntryLog }
+          .compactMap { String($0.composedMessage.suffix(from: $0.composedMessage.index($0.composedMessage.startIndex, offsetBy: 2)))
+          }
+    } catch {
+      // If an error happened reading the OSLogStore, surface the error as an UpdatesLogEntry with the right shape
```
[the remark this answers is not in the document]
Comment (douglowder) [reply packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift:58]: Ok modified this to throw instead, and made appropriate adjustments to test code

File: packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogEntry.swift [hunk 24]
```diff
@@ -0,0 +1,77 @@
...
+    result["updateId"] = updateId != nil ? updateId : NSNull()
+    result["assetId"] = assetId != nil ? assetId : NSNull()
+    if stacktrace != nil {
+      let nsstack = NSMutableArray()
+      for s in stacktrace! {
+        nsstack.add(s)
+      }
+      result["stacktrace"] = nsstack
+    } else {
+      result["stacktrace"] = NSNull()
```
Comment (wschurman) [inline packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogEntry.swift:48]: I believe we can use swift types here, and they'll be automatically converted over the bridge (for free)

**Review by wschurman — APPROVED** [review]
One more small comment but otherwise looks good!

File: packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogEntry.swift (hunk 24 above)
Comment (douglowder) [reply packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogEntry.swift:48]: Sadly this does not seem to be the case. Passing in `nil` in Swift does not show up as an NSNull in OC.

Perhaps a cleaner way is to simply not have the keys present in the dictionary for nulls.

Comment (douglowder) [reply packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogEntry.swift]: Fixed

File: packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogEntry.swift [hunk 25]
```diff
@@ -0,0 +1,79 @@
...
+    result["timestamp"] = timestamp
+    result["message"] = message
+    result["code"] = code
+    result["level"] = level
+    if (updateId != nil) {
+      result["updateId"] = updateId
+    }
+    if (assetId != nil) {
+      result["assetId"] = assetId
+    }
```
Comment (tsapeta) [inline packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogEntry.swift:44]: Generally, we don't use `()` in ifs, switches, etc.
Is there any reason why you check if it's non-nil? I think you can just assign it and don't care 😅

File: packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogEntry.swift [hunk 26]
```diff
@@ -0,0 +1,79 @@
...
+    if stacktrace != nil {
+      let nsstack = NSMutableArray()
```
Comment (tsapeta) [inline packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogEntry.swift:46]: Why `NSMutableArray`? Please use Swift types.

File: packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogEntry.swift [hunk 27]
```diff
@@ -0,0 +1,79 @@
...
+      for s in stacktrace! {
```
Comment (tsapeta) [inline packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogEntry.swift:47]: I think I already mentioned this somewhere — never use force unwrapping `!` except in tests and some edge-cases where it makes sense and we're sure it won't crash. Here the situation is simple, just do
```
if let stacktrace = stacktrace {
```
instead of
```
if stacktrace != nil {
```
to safely unwrap the variable

File: packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogEntry.swift [hunk 28]
```diff
@@ -0,0 +1,79 @@
...
+      return try jsonDecoder.decode(UpdatesLogEntry.self, from: jsonData)
+    } catch {
+      return nil
+    }
+  }
+
+  /**
+   Utility method to construct stacktrace as a string array for log entries
+   */
+  private static let STACKTRACE_MAX_LENGTH = 20
```
Comment (tsapeta) [inline packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogEntry.swift:71]: I guess you meant to put this comment a few lines below to refer to `currentStackTrace` instead of the `STACKTRACE_MAX_LENGTH` constant 😛

File: packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift [hunk 29]
```diff
@@ -0,0 +1,54 @@
...
+    // Fetch log objects, selecting our subsystem and category
+    let predicate = NSPredicate(format: "category == %@ AND subsystem = %@",
+                                argumentArray: [UpdatesLogger.EXPO_UPDATES_LOG_CATEGORY, Logger.EXPO_MODULES_LOG_SUBSYSTEM])
+    let allEntries = try logStore.getEntries(at: position, matching: predicate)
+
+    // Extract just the log message strings, removing the first two characters added
+    // by ExpoModulesCore.Logger
+    return allEntries
+        .compactMap { String($0.composedMessage.suffix(from: $0.composedMessage.index($0.composedMessage.startIndex, offsetBy: 2)))
+        }
```
Comment (tsapeta) [inline packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift:52]:
```suggestion
    return allEntries
      .map { entry in
        let suffixFrom = entry.composedMessage.index(entry.composedMessage.startIndex, offsetBy: 2)
        return String(entry.composedMessage.suffix(from: suffixFrom))
      }
```

**Review by tsapeta — CHANGES_REQUESTED** [review]

File: packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogEntry.swift (hunk 25 above)
Comment (douglowder) [reply packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogEntry.swift:44]: I did this to ensure that OC code got the dictionaries with the right behavior. I'll check to see if that is still necessary.

Comment (tsapeta) [reply packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogEntry.swift:44]: Since this is Swift dictionary, assigning `nil` means nothing else than just deleting the key from the dictionary if it existed

File: packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogEntry.swift (hunk 27 above)
Comment (douglowder) [reply packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogEntry.swift:47]: That is an interesting syntax

File: packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogEntry.swift (hunk 25 above)
Comment (douglowder) [reply packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogEntry.swift:44]: Fixed the syntax

File: packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogEntry.swift (hunk 27 above)
Comment (douglowder) [reply packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogEntry.swift:47]: Fixed

File: packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogEntry.swift (hunk 28 above)
Comment (douglowder) [reply packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogEntry.swift:71]: Fixed

File: packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift (hunk 29 above)
Comment (douglowder) [reply packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogReader.swift:52]: Fixed

File: packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogEntry.swift (hunk 26 above)
Comment (douglowder) [reply packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogEntry.swift:46]: Fixed

File: packages/expo-updates/ios/EXUpdates/Logging/UpdatesErrorCode.swift [hunk 30]
```diff
@@ -0,0 +1,43 @@
...
+  case UpdateHasInvalidSignature = 4
+  case UpdateFailedToLoad = 5
+  case AssetsFailedToLoad = 6
+  case JSRuntimeError = 7
+
+  // Because this enum is exported to Objective-C,
+  // the usual "\(UpdatesErrorCode.NoUpdatesAvailable)"
+  // string representation will not work as expected,
+  // so we add this representation here
+  public var asString: String {
```
Comment (tsapeta) [inline packages/expo-updates/ios/EXUpdates/Logging/UpdatesErrorCode.swift:23]: I've got an idea, maybe we can make the enum be a String instead of Int? Then I think we won't have to map it to string 🤔

File: packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogEntry.swift [hunk 31]
```diff
@@ -0,0 +1,79 @@
...
+    }
+    if let assetId = assetId {
+      result["assetId"] = assetId
+    }
+    if let stacktrace = stacktrace {
+      var nsstack = [String]()
+      for s in stacktrace {
+        nsstack.append(s)
+      }
+      result["stacktrace"] = nsstack
```
Comment (tsapeta) [inline packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogEntry.swift:50]: Isn't it the same? 😉 
```suggestion
      result["stacktrace"] = stacktrace
```

File: packages/expo-updates/CHANGELOG.md [hunk 32]
```diff
@@ -6,6 +6,8 @@
 
 ### 🎉 New features
 
+- Logging and log reading. ([#18284](https://github.com/expo/expo/pull/18284) by [@douglowder](https://github.com/douglowder))
```
Comment (tsapeta) [inline packages/expo-updates/CHANGELOG.md:9]: Could you make it more descriptive? Also, it's not a new feature from the user perspective — it doesn't change the public API and doesn't bring any visible value for the user, so I would prefer moving this to "Others" category.
This reminded me that I should clarify it in our contributing guide about updating changelogs 😅

File: packages/expo-updates/ios/EXUpdates/Logging/UpdatesErrorCode.swift (hunk 30 above)
Comment (douglowder) [reply packages/expo-updates/ios/EXUpdates/Logging/UpdatesErrorCode.swift:23]: Funny story, that's exactly what I did on Android. However, I think it's better this way here as an integer enum is more usual in Objective C.

File: packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogEntry.swift (hunk 31 above)
Comment (douglowder) [reply packages/expo-updates/ios/EXUpdates/Logging/UpdatesLogEntry.swift:50]: Ah yes, it was not the same when I was doing NSArray. I'll fix this.

File: packages/expo-updates/CHANGELOG.md (hunk 32 above)
Comment (douglowder) [reply packages/expo-updates/CHANGELOG.md:9]: True -- and the expotools add-changelog should support "Others" as well :)

**Review by tsapeta — APPROVED** [review]
